AZ3166/src/libraries/WebSocket/src/WebSocketClient.cpp

```diff
@@ -0,0 +1,448 @@
+    if (portPtr != NULL)
+    {
+        hostLen = portPtr - hostPtr;
+        portPtr++;
+        if (sscanf(portPtr, "%hu", port) != 1)
+        {
+            Serial.println("ERROR: Could not find port.");
+            return -1;
+        }
+    }
+    else
+    {
+        *port = 0;
+    }
+    char *pathPtr = strchr(hostPtr, '/');
+    if (pathPtr == NULL)
+    {
+        Serial.println("ERROR: Path not specified. Please add /[path] to the end of the websocket address");
+        return -1;
+    }
+    if (hostLen == 0)
+    {
+        hostLen = pathPtr - hostPtr;
+    }
+
+    if (maxHostLen < hostLen + 1) //including NULL-terminating char
+    {
+        Serial.printf("ERROR: Host str is too small (%d >= %d)", maxHostLen, hostLen + 1);
+        return -1;
+    }
+    memcpy(host, hostPtr, hostLen);
+    host[hostLen] = '\0';
+
+    size_t pathLen;
+    char *fragmentPtr = strchr(hostPtr, '#');
+    if (fragmentPtr != NULL)
+    {
+        pathLen = fragmentPtr - pathPtr;
+    }
+    else
+    {
+        pathLen = strlen(pathPtr);
+    }
+
+    if (maxPathLen < pathLen + 1) //including NULL-terminating char
+    {
+        Serial.printf("ERROR: Path str is too small (%d >= %d)", maxPathLen, pathLen + 1);
+        return -1;
+    }
+    memcpy(path, pathPtr, pathLen);
+    path[pathLen] = '\0';
+
+    return 0;
+}
+
+bool WebSocketClient::connect()
+{
+    if (_tcpSocket != NULL)
+    {
+        // Already connected.
+        return true;
+    }
+    else
+    {
+        _tcpSocket = new TCPSocket();
+        if (_tcpSocket == NULL)
+        {
+            return false;
+        }
+
+        if (_tcpSocket->open(WiFiInterface()) != 0 || _tcpSocket->connect(host, port) != 0)
+        {
+            Serial.printf("ERROR: Unable to connect to (%s) on port (%d)\r\n", host, port);
+            wait(0.2);
+
+            delete _tcpSocket;
+            _tcpSocket = NULL;
+            return false;
+        }
+
+        _tcpSocket->set_timeout(1000);
+    }
+
+    char cmd[200];
+
+    // sent http header to upgrade to the ws protocol
+    sprintf(cmd, "GET %s HTTP/1.1\r\n", path);
+    write(cmd, strlen(cmd));
+
+    sprintf(cmd, "Host: %s:%d\r\n", host, port);
+    write(cmd, strlen(cmd));
+
+    sprintf(cmd, "Upgrade: WebSocket\r\n");
+    write(cmd, strlen(cmd));
+
+    sprintf(cmd, "Connection: Upgrade\r\n");
+    write(cmd, strlen(cmd));
+
+    sprintf(cmd, "Sec-WebSocket-Key: L159VM0TWUzyDxwJEIEzjw==\r\n");
+    write(cmd, strlen(cmd));
+
+    sprintf(cmd, "Sec-WebSocket-Version: 13\r\n\r\n");
+    int ret = write(cmd, strlen(cmd));
+    if (ret != strlen(cmd))
+    {
+        close();
+        Serial.println("ERROR: Could not send request.");
+        return false;
+    }
+
+    ret = read(cmd, 200, 100);
+    if (ret < 0)
+    {
+        close();
+        Serial.println("ERROR: Could not receive answer.");
+        return false;
+    }
+
+    cmd[ret] = '\0';
+
+    if (strstr(cmd, "DdLWT/1JcX+nQFHebYP+rqEx5xI=") == NULL)

```

Comment:
stupid question, this magic string is part of the protocol ?

Reply:
Yes ! Magic !